### PR TITLE
IGNITE-12614 : Disallow silent deactivation of cluster to prevent in-mem data loss.

### PR DESCRIPTION
**Pull request by Vladsz83**
First implementation for conceptual review.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 1]
```diff
@@ -3946,7 +3952,27 @@ private void checkNearCacheStarted(IgniteCacheProxy<?, ?> cache) throws IgniteCh
...
+
+    /** {@inheritDoc} */
+    @Override public void activate() {
+        cluster().state(ClusterState.ACTIVE);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void deactivate(boolean force) {
+        //Check if cluster ir ready for deactivation.
+        if (cluster().state() == ClusterState.ACTIVE && !force) {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Cluster state can be `ACTIVE_READ_ONLY`

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java [hunk 2]
```diff
@@ -388,11 +388,31 @@
     )
     public boolean pingNode(String nodeId);
 
+    /** Activates cluster. */
+    @MXBeanDescription(
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Seems we can make one-line them.

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java [hunk 3]
```diff
@@ -388,11 +388,31 @@
...
+    public void deactivate(boolean force);
+
     /**
      * @param active Activate/DeActivate flag.
+     * @deprecated Use {@link #activate()} and {@link #deactivate(boolean)} instead.
      */
+    @Deprecated
     @MXBeanDescription(
-        "Execute activate or deactivate process."
+        "Execute activate or deactivate process. Deprecated. Use activate() / deactivate(). Deactivation may require flag 'force'."
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: More than 120 chars in one line

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 4]
```diff
@@ -102,4 +167,53 @@
...
+        @SuppressWarnings("unchecked")
+        @Override public Boolean execute() throws IgniteException {
+            //Find any node with disabled memory reusage on deactivation/activation.
+            boolean cacheDataCanBeLost = ignite.cluster().forPredicate(node->
+                (Boolean)node.attributes().getOrDefault(ATTR_MEMORY_ERASURE_ON_DEACTIVATION, true))
+                .nodes().stream().findAny().isPresent();
+
+            if (cacheDataCanBeLost) {
+                //Find data region to set persistent flag.
+                for (String cacheName : ignite.cacheNames()) {
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Lets use the same approach as for JMX

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 5]
```diff
@@ -42,6 +63,26 @@
...
+    /** Force cluster deactivation even it might have in-mem caches. */
+    private boolean force;
+
+    /**
+     * Checks if it is resonable to deactivate cluster.
+     *
+     * @param taskLauncher Computation task launcher. The task has no params, just needs to be launched for its result.
+     * @return Empty (not-null) message if cluster is ready. Or warning text telling why deactivation is not advised.
+     */
+    @NotNull public static String isClusterReadyForDeactivation(
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Not actual. Code changed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 6]
```diff
@@ -1959,6 +1962,9 @@ private void fillNodeAttributes(boolean notifyEnabled) throws IgniteCheckedExcep
 
         ctx.addNodeAttribute(ATTR_EVENT_DRIVEN_SERVICE_PROCESSOR_ENABLED,
             ctx.service() instanceof IgniteServiceProcessor);
+
+        //Allows to predict behavior on deactivation.
+        add(ATTR_MEMORY_ERASURE_ON_DEACTIVATION, !getBoolean(IGNITE_REUSE_MEMORY_ON_DEACTIVATE));
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 7]
```diff
@@ -3946,7 +3952,27 @@ private void checkNearCacheStarted(IgniteCacheProxy<?, ?> cache) throws IgniteCh
 
     /** {@inheritDoc} */
     @Override public void active(boolean active) {
-        cluster().active(active);
+        if (active)
+            activate();
+        else
+            deactivate(false);
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 1 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 8]
```diff
@@ -3946,7 +3952,27 @@ private void checkNearCacheStarted(IgniteCacheProxy<?, ?> cache) throws IgniteCh
...
+    /** {@inheritDoc} */
+    @Override public void activate() {
+        cluster().state(ClusterState.ACTIVE);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void deactivate(boolean force) {
+        //Check if cluster ir ready for deactivation.
+        if (cluster().state() == ClusterState.ACTIVE && !force) {
+            String msg = ClusterStateChangeCommand.isClusterReadyForDeactivation((cls -> compute().execute(cls, null)));
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Not actual. Code changed.

[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Implemented.

[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Implemented.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 9]
```diff
@@ -68,14 +110,29 @@
     /** {@inheritDoc} */
     @Override public Object execute(GridClientConfiguration clientCfg, Logger log) throws Exception {
         try (GridClient client = Command.startClient(clientCfg)) {
+
+            //Search for in-memory-only caches. Fail if possible data loss.
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Fixed. Checked other similar.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 10]
```diff
@@ -91,6 +148,14 @@
         catch (IllegalArgumentException e) {
             throw new IllegalArgumentException("Can't parse new cluster state. State: " + s, e);
         }
+
+        if (argIter.hasNextArg()) {
+            String arg = argIter.peekNextArg();
+            if ("--force".equalsIgnoreCase(arg)) {
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 11]
```diff
@@ -102,4 +167,53 @@
     @Override public String name() {
         return SET_STATE.toCommandName();
     }
+
+    /** Searches for any non-persistent cache. */
+    private static class FindNotPersistentCachesJob extends ComputeJobAdapter {
+
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Not actual. Code changed.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java (hunk 4 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Implemented.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java [hunk 12]
```diff
@@ -59,25 +69,51 @@
...
         }
 
         return null;
     }
 
+    /** {@inheritDoc} */
+    @Override public void parseArguments(CommandArgIterator argIter) {
+        if (argIter.hasNextArg()) {
+            String arg = argIter.peekNextArg();
+            if ("--force".equalsIgnoreCase(arg)) {
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java [hunk 13]
```diff
@@ -59,25 +65,51 @@
...
+                force = true;
+                argIter.nextArg("");
```
[the remark this answers is not in the document]
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Fixed. Also the others.

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 3 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java [hunk 14]
```diff
@@ -33,12 +34,19 @@
  */
 @Deprecated
 public class DeactivateCommand implements Command<Void> {
+    /**  */
+    private static final String FLAG_FORCE = "--force";
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/DeactivateCommand.java]: Lets reuse var from the ClusterStateChangeCommand class

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java [hunk 15]
```diff
@@ -424,12 +427,8 @@
      * @throws JMException Thrown if undeploy failed.
      */
     @MXBeanDescription("Makes the best attempt to undeploy a task from the whole grid.")
-    @MXBeanParametersNames(
-        "taskName"
-    )
-    @MXBeanParametersDescriptions(
-        "Name of the task to undeploy."
-    )
+    @MXBeanParametersNames("taskName")
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Lets do not change code that is not related to the issue

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java [hunk 16]
```diff
@@ -3949,6 +3957,28 @@ private void checkNearCacheStarted(IgniteCacheProxy<?, ?> cache) throws IgniteCh
...
 
+    /** {@inheritDoc} */
+    @Override public void activate() {
+        cluster().state(ClusterState.ACTIVE);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void deactivate(boolean force) {
+        // Check if cluster ir ready for deactivation.
+        if ((cluster().state() == ClusterState.ACTIVE || cluster().state() == ClusterState.ACTIVE_READ_ONLY)
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: I suggest:
1. Change condition to `cluster().state() != ClusterState.INACTIVE && !force`
2. Do not execute task. Implement some method like this `ctx.state().isCacheDataCanBeLostOnDeactivation()` which will check attributes and in-memry caches

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java [hunk 17]
```diff
@@ -36,12 +37,18 @@
  * Command to change cluster state.
  */
 public class ClusterStateChangeCommand implements Command<ClusterState> {
+    /**  */
+    private static final String FLAG_FORCE = "--force";
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: I suggest name it `FORCE_COMMAND`

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java [hunk 18]
```diff
@@ -380,23 +376,30 @@
...
     public boolean pingNode(String nodeId);
 
+    /** Activates cluster. */
+    @MXBeanDescription("Execute activation process.")
+    @MXBeanParametersNames("active")
+    public void activate();
+
+    /** Deactivates cluster. */
+    @MXBeanDescription("Execute deactivation process.")
+    @MXBeanParametersNames("force")
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: We should add MXBeanParametersDescriptions

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java [hunk 19]
```diff
@@ -380,23 +376,30 @@
...
-    @MXBeanParametersDescriptions(
-        "String presentation of node ID. See java.util.UUID class for details."
-    )
+    @MXBeanParametersNames("nodeId")
+    @MXBeanParametersDescriptions("String presentation of node ID. See java.util.UUID class for details.")
     public boolean pingNode(String nodeId);
 
+    /** Activates cluster. */
+    @MXBeanDescription("Execute activation process.")
+    @MXBeanParametersNames("active")
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: There is no parametrs

File: modules/core/src/main/java/org/apache/ignite/internal/visor/cluster/VisorCheckDeactivationTask.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Task for checking if deactiation is safe. Returns false when deactivation can lead to data loss.
+ */
+@GridInternal
+public class VisorCheckDeactivationTask extends VisorOneNodeTask<Void, Boolean> {
+    /** Serial version uid. */
+    private static final long serialVersionUID = 0L;
+
+    /** */
+    public static final String WARN_DEACTIVATION_IN_MEM_CACHES =
```
Comment (NSAmelchev) [inline modules/core/src/main/java/org/apache/ignite/internal/visor/cluster/VisorCheckDeactivationTask.java]: I think we can move it to the GridClusterStateProcessor. Add javadoc, please

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 15 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java (hunk 16 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/IgniteKernal.java]: Impelemted

File: modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java (hunk 17 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/commandline/ClusterStateChangeCommand.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 18 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Fixed

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 19 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: Yep. Fixed.

File: modules/core/src/main/java/org/apache/ignite/internal/visor/cluster/VisorCheckDeactivationTask.java (hunk 20 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/visor/cluster/VisorCheckDeactivationTask.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 18 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: @nizhikov , please review IgniteMXBean. We have to put flag 'force' somewhere for the deactivation. Current choice is

`IgniteMXBean#clusterState(String state) -> clusterState(String state, boolean force)`

Usually one has to support backward compatibility of an Interface. But this is an JMX. I don't think it is unsafe to add single param. Otherwise we need to implement additional method like

` clusterState(String state, boolean force)
or
deactivate(boolean force)`

whereas the mxbean already has several methods for the cluster state:
`
void active(boolean active)
boolean active()
void clusterState(String state)
String clusterState()
`

Don't think one more would be handy;

Also, should we declare 
`IgniteMXBean#active(boolean active)`
deprecated too.

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: > But this is an JMX. I don't think it is unsafe to add single param. 

Why do you think so?

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: > > But this is an JMX. I don't think it is unsafe to add single param.
> 
> Why do you think so?

An automated, external system like Zabbix of something must use exposed bean to be afraid of it. Manual method invocation in the console is quite obvious. Just requires to say extra ‘yes’ or ‘no’ for the new argument. Direct bean usage within the code isn't reasonable except the tests.

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: JMX bean is a public API, so we should preserve backward compatibility.

Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: > JMX bean is a public API, so we should preserve backward compatibility.

Ok.

And

>> Also, should we declare
>> `IgniteMXBean#active(boolean active)`
>> deprecated too?

Comment (nizhikov) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: I think it should be discussed on the dev-list

File: modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/mxbean/IgniteMXBean.java]: @nizhikov , how do you think, can we keep IgniteMXBean as it is now in this PR? With one new 

`clusterState(String state, boolean force)`

Unfortunately, we can't use two different default-methods in the interfaces IgniteMXBean and Ignite:
`    
default public void active(boolean active){
        active(active, false);
    }

and
default public void active(boolean active){
    active(active, true);
}
`

Jaca compilator forces us to choose one in IgniteKernal:
`class org.apache.ignite.internal.IgniteKernal inherits unrelated defaults for active(boolean) from types org.apache.ignite.Ignite and org.apache.ignite.mxbean.IgniteMXBean`

I think this happens because there are no real implementations of default methods within interfaces. They are just code sample for in-lining to the implementation. So, we have two different implementations.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java [hunk 21]
```diff
@@ -1654,6 +1659,14 @@ public boolean isBaselineAutoAdjustEnabled() {
         return distributedBaselineConfiguration.isBaselineAutoAdjustEnabled();
     }
 
+    /**
+     * @return {@code False} if cluster deactivation can erase user's data and objects.
+     * {@code True} If deactivation is safe.
+     */
+    public boolean isDeactivationSafe() {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cluster/GridClusterStateProcessor.java]: Please, use `ctx.cache().cacheDescriptors()` here and remove `GridCacheProcessor#inMemoryCaches` method.

```
        return ctx.cache().cacheDescriptors().values().stream()
            .filter(desc -> !isPersistentCache(desc.cacheConfiguration(), ctx.config().getDataStorageConfiguration()))
            .findFirst()
            .isPresent();
```
